src/api/spec/models/project_spec.rb

```diff
@@ -287,4 +287,99 @@
+      allow_any_instance_of(ProjectFile).to receive(:to_s).with(deleted: 1).and_return(
+        "<revisionlist>\n  <revision rev=\"1\" vrev=\"\">\n    <srcmd5>d41d8cd98f00b204e9800998ecf8427e</srcmd5>\n    " \
+        "<version></version>\n    <time>1498113679</time>\n    <user>Admin</user>\n    <comment>1</comment>\n  " \
+        "</revision>\n</revisionlist>\n"
+      )
+
+      expect(Project.deleted?('very-nice-project-name')).to be_truthy
+    end
+  end
+
+  describe '.restore' do
+    let(:admin_user) { create(:admin_user, login: 'Admin') }
+    let(:deleted_project) {
+      create(
+        :project_with_packages,
+        {
+          name:                'project_used_for_restoration',
+          title:               'restoration_project_title',
+          package_title:       'restoration_title',
+          package_description: 'restoration_desc',
+          package_name:        'restoration_package'
+        }
+      )
+    }
+
+    before do
+      login admin_user
+      deleted_project.destroy!
+    end
+
+    describe 'history element' do
+      it 'sets the user that restored the project' do
+        Project.restore(deleted_project.name, user: admin_user.login)
+
+        meta = Xmlhash.parse(ProjectFile.new(project_name: deleted_project.name, name: '_history').to_s(deleted: 1))
+        expect(meta['revision'].last['user']).to eq(admin_user.login)
+      end
+    end
+
+    describe 'project meta' do
+      let(:project_meta_xml) { ProjectMetaFile.new(project_name: deleted_project.name).to_s }
+
+      it 'gets properly updated' do
+        restored_project = Project.restore(deleted_project.name)
+        expect(ProjectMetaFile.new(project_name: restored_project.name).to_s).to eq(project_meta_xml)
+      end
+    end
+
+    describe 'restoration of project with packages' do
+      let(:package1) { deleted_project.packages.first }
+      let(:package1_meta_before_deletion) { package1.render_xml }
+      let(:package2) { deleted_project.packages.last }
+      let(:package2_meta_before_deletion) { package2.render_xml }
+
+      before do
+        # for some reason, Collect.find finds always 5 packages for `deleted_project`. Even if
+        # you use a different factory where a project doesn't have any packages before.
+        # That's why we need to mock it...
+        allow(Collection).to receive(:find).and_return(
+          ActiveXML::Node.new(
+            "<collection>#{package1_meta_before_deletion}#{package2_meta_before_deletion}</collection>"
+          )
+        )
+      end
+
+      it 'creates package records in the database' do
+        restored_project = Project.restore('project_used_for_restoration')
+        expect(restored_project.packages.size).to eq(2)
+      end
+
+      it 'verifys the meta of restored packages' do
+        Project.restore('project_used_for_restoration')
+
+        expect(Package.find_by(name: package1.name).render_xml).to eq(package1_meta_before_deletion)
```

Comment:
Is better to have the restored project in a variable and find the packages in the relationship... instead of using Package.find_by